### ISSUE TITLE
Investigate amino performance/mem usage

### ISSUE THREAD

**Issue by liamsi** [tendermint/go-amino#254]
While looking into adding amino to this list: https://github.com/alecthomas/go_serialization_benchmarks (see this branch: https://github.com/Liamsi/go_serialization_benchmarks/tree/add_amino), it seems like amino is quite slow compared to other similar libraries:
```
BenchmarkAminoMarshal-12                         300000              5492 ns/op            4138 B/op         72 allocs/op
BenchmarkAminoUnmarshal-12                       1000000              1081 ns/op             194 B/op          7 allocs/op
BenchmarkProtobufMarshal-12                      2000000               676 ns/op             200 B/op          7 allocs/op
BenchmarkProtobufUnmarshal-12                    3000000               593 ns/op             192 B/op         10 allocs/op
BenchmarkGoprotobufMarshal-12                    5000000               325 ns/op              96 B/op          2 allocs/op
BenchmarkGoprotobufUnmarshal-12                  3000000               498 ns/op             200 B/op         10 allocs/op
BenchmarkGogoprotobufMarshal-12                 20000000               119 ns/op              64 B/op          1 allocs/op
BenchmarkGogoprotobufUnmarshal-12               10000000               166 ns/op              96 B/op          3 allocs/op
```

Hopefully, we should be able to vastly improve this performance without completely reworking the structure. 

Here is what the profiler tells us about memory/cpu while running above benchmarks for unmarshaling/marshaling:

--------
--------
![marshal_cpu_profile](https://user-images.githubusercontent.com/7441766/53746496-93b0b900-3ea1-11e9-9af8-cf9f7502ec21.png)
--------
--------
![marshal_mem_profile](https://user-images.githubusercontent.com/7441766/53746497-94494f80-3ea1-11e9-9354-d190e40ddb91.png)
--------
--------
![unmarshal_cpu_profile](https://user-images.githubusercontent.com/7441766/53746498-94494f80-3ea1-11e9-84bb-e1f74df0ef97.png)
--------
--------
![unmarshal_mem_profile](https://user-images.githubusercontent.com/7441766/53746500-94494f80-3ea1-11e9-9562-77c5fdd85a09.png)
--------
--------

Zaki suggested to add a hint in the readme about performance issue and in which cases users would might want to refrain from using amino.

**Comment by rickyyangz** [comment]
Two things are not fair for your amino case. 
1. Before marshalling, you need to reset timer or stop timer before generating testing data. 
```go
func BenchmarkAminoMarshal(b *testing.B) {
        // b.StopTimer()
	data := generateAmino()
	b.ReportAllocs()
	b.StartTimer() 
	s := AminoSerializer{amino.NewCodec()}
	for i := 0; i < b.N; i++ {
			s.MustMarshalBinaryBare(data[rand.Intn(len(data))])
	}
}
```

2. the type of `BirthDay` is time.Time while in other cases it is int64

**Comment by liamsi** [comment]
good point @rickyyangz. Did you you re-benchmark with your suggested changes? I would assume the performance to still be much slower than generated protobuf.
